File: 5.最长回文子串.ts
/*
 * @Author: 彭江
 * @Date: 2022-04-25 20:29:19
 * @LastEditors: 彭江
 * @LastEditTime: 2022-04-25 23:32:12
 */
/*
 * @lc app=leetcode.cn id=5 lang=typescript
 *
 * [5] 最长回文子串
 */

// @lc code=start


function longestPalindrome(s: string): string {
  let longAnswer: string = '';
  const n = s.length;

  if (!s || n === 1) {
    longAnswer = s;
  }

  const expond = (l: number, r: number) => {
    while (l >= 0 && r <= n - 1 && s[l] === s[r]) {
      l--;
      r++;
    }
    return s.slice(l + 1, r);
  }

  for (let i = 0; i < n - 1; i++) {
    const l1 = expond(i, i);
    const l2 = expond(i, i+1);
    const arr = [l1,l2];
    arr.forEach(str => {
      if (str.length > longAnswer.length) {
        longAnswer = str;
      }
    })
  }

  return longAnswer;
};

// longestPalindrome("babad");
// @lc code=end

File: 5.最长回文子串.go
package leetcode

/*
 * @lc app=leetcode.cn id=5 lang=golang
 *
 * [5] 最长回文子串
 */

// @lc code=start

func longestPalindrome(s string) string {
	var longAnswer string = ""
	var n int = len(s)

	if n <= 1 {
		longAnswer = s
	}

	var expond = func(l int, r int) string {
		for l >= 0 && r <= n-1 && s[l] == s[r] {
			l--
			r++
		}
		return s[l+1 : r]
	}

	for i := 0; i < n-1; i++ {
		l1 := expond(i, i)
		l2 := expond(i, i+1)
		var arr [2]string = [2]string{l1, l2}
		for j := 0; j < len(arr); j++ {
			val := arr[j]
			if len(val) > len(longAnswer) {
				longAnswer = val
			}
		}
	}

	return longAnswer
}

// func main() {
// 	var pa string = "ajgiljtperkvubjmdsefcylksrxtftqrehoitdgdtttswwttmfuvwgwrruuqmxttzsbmuhgfaoueumvbhajqsaxkkihjwevzzedizmrsmpxqavyryklbotwzngxscvyuqjkkaotitddlhhnutmotupwuwyltebtsdfssbwayuxrbgihmtphshdslktvsjadaykyjivbzhwujcdvzdxxfiixnzrmusqvwujjmxhbqbdpauacnzojnzxxgrkmupadfcsujkcwajsgintahwgbjnvjqubcxajdyyapposrkpqtpqfjcvbhlmwfutgognqxgaukpmdyaxghgoqkqnigcllachmwzrazwhpppmsodvxilrccfqgpkmdqhoorxpyjsrtbeeidsinpeyxxpsjnymxkouskyhenzgieybwkgzrhhrzgkwbyeigznehyksuokxmynjspxxyepnisdieebtrsjypxroohqdmkpgqfccrlixvdosmppphwzarzwmhcallcginqkqoghgxaydmpkuagxqngogtufwmlhbvcjfqptqpkrsoppayydjaxcbuqjvnjbgwhatnigsjawckjuscfdapumkrgxxznjozncauapdbqbhxmjjuwvqsumrznxiifxxdzvdcjuwhzbvijykyadajsvtklsdhshptmhigbrxuyawbssfdstbetlywuwputomtunhhlddtitoakkjquyvcsxgnzwtoblkyryvaqxpmsrmzidezzvewjhikkxasqjahbvmueuoafghumbszttxmquurrwgwvufmttwwstttdgdtioherqtftxrsklycfesdmjbuvkreptjligja"
// 	longestPalindrome(pa)
// }

// @lc code=end
